app/scripts/controllers/transactions/tx-state-manager.js

```diff
@@ -187,28 +187,43 @@ export default class TransactionStateManager extends EventEmitter {
     const transactions = this.getTransactions({
       filterToCurrentNetwork: false,
     });
-    const txCount = transactions.length;
     const { txHistoryLimit } = this;
 
     // checks if the length of the tx history is longer then desired persistence
     // limit and then if it is removes the oldest confirmed or rejected tx.
     // Pending or unapproved transactions will not be removed by this
-    // operation.
+    // operation. For safety of presenting a fully functional transaction UI
+    // representation, this function will not break apart transactions with the
+    // same nonce, per network. Not accounting for transactions of the same
+    // nonce and network combo can result in confusing or broken experiences
+    // in the UI.
     //
     // TODO: we are already limiting what we send to the UI, and in the future
     // we will send UI only collected groups of transactions *per page* so at
     // some point in the future, this persistence limit can be adjusted. When
     // we do that I think we should figure out a better storage solution for
     // transaction history entries.
-    if (txCount > txHistoryLimit - 1) {
-      const index = transactions.findIndex((metaTx) => {
-        return getFinalStates().includes(metaTx.status);
-      });
-      if (index !== -1) {
-        this._deleteTransaction(transactions[index].id);
-      }
-    }
+    const nonceNetworkSet = new Set();
+    const txsToDelete = transactions
+      .reverse()
+      .filter((tx) => {
+        const { nonce } = tx.txParams;
+        const { chainId, metamaskNetworkId, status } = tx;
+        const key = `${nonce}-${chainId ?? metamaskNetworkId}`;
+        if (nonceNetworkSet.has(key)) {
+          return false;
+        } else if (
+          nonceNetworkSet.size < txHistoryLimit - 1 ||
+          getFinalStates().includes(status) === false
```

Comment:
I'm not sure about this :thinking: This dictates that a transaction in its final state won't count towards the limit. But it should count - I don't see why it wouldn't.

Reply:
its transactions NOT in their final state, AKA unapproved/pending txs

Reply:
Yes, but that's still wrong isn't it?

e.g. if we have 500 transactions with unique nonces that are all completed, this would skip over all of them, removing nothing.

Reply:
Oh nevermind, my bad. it's OR not and. I get it now.